uniffi_meta/src/types.rs

```diff
@@ -0,0 +1,147 @@
+//! bindings. Think "objects" and "enums" and "records".
+//!
+//! The [`Type`] enum represents high-level types that would appear in the public API of
+//! a component, such as enums and records as well as primitives like ints and strings.
+//! The Rust code that implements a component, and the foreign language bindings that consume it,
+//! will both typically deal with such types as their core concern.
+//!
+//! As a developer working on UniFFI itself, you're likely to spend a fair bit of time thinking
+//! about how these API-level types map into the lower-level types of the FFI layer as represented
+//! by the [`ffi::FfiType`](super::ffi::FfiType) enum, but that's a detail that is invisible to end users.
+
+use serde::Serialize;
+
+use crate::Checksum;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Checksum, Ord, PartialOrd, Serialize)]
+pub enum ObjectImpl {
+    Struct,
+    Trait,
+}
+
+impl ObjectImpl {
+    /// Return the fully qualified name which should be used by Rust code for
+    /// an object with the given name.
+    /// Includes `r#`, traits get a leading `dyn`. If we ever supported associated types, then
+    /// this would also include them.
+    pub fn rust_name_for(&self, name: &str) -> String {
+        if self == &ObjectImpl::Trait {
+            format!("dyn r#{name}")
+        } else {
+            format!("r#{name}")
+        }
+    }
+
+    // uniffi_meta and procmacro support tend to carry around `is_trait` bools. This makes that
+    // mildly less painful
+    pub fn from_is_trait(is_trait: bool) -> Self {
+        if is_trait {
+            ObjectImpl::Trait
+        } else {
+            ObjectImpl::Struct
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Checksum, Ord, PartialOrd, Serialize)]
+pub enum ExternalKind {
+    Interface,
+    // Either a record or enum
+    DataClass,
+}
+
+/// Represents all the different high-level types that can be used in a component interface.
+/// At this level we identify user-defined types by name, without knowing any details
+/// of their internal structure apart from what type of thing they are (record, enum, etc).
+#[derive(Debug, Clone, Eq, PartialEq, Checksum, Ord, PartialOrd, Serialize)]
+pub enum Type {
+    // Primitive types.
+    UInt8,
+    Int8,
+    UInt16,
+    Int16,
+    UInt32,
+    Int32,
+    UInt64,
+    Int64,
+    Float32,
+    Float64,
+    Boolean,
+    String,
+    Bytes,
+    Timestamp,
+    Duration,
+    Object {
+        // The name in the "type universe"
+        name: String,
+        // How the object is implemented.
+        imp: ObjectImpl,
+    },
+    ForeignExecutor,
+    // Types defined in the component API, each of which has a string name.
+    Record(String),
+    Enum(String),
+    CallbackInterface(String),
+    // Structurally recursive types.
+    Optional(Box<Type>),
+    Sequence(Box<Type>),
+    Map(Box<Type>, Box<Type>),
```

Comment:
Can this be `Map { key_type: Box<Type>, value_type: Box<Type> }`?

I don't mind the tuple-style for Optional and Sequence, but once we get multiple child types I think named fields are better.

Reply:
sure! How about just `k` and `v`?